src/adm/filter.rs

```diff
@@ -182,44 +183,40 @@ impl AdmFilter {
-                .read(&host, bucket.path().trim_start_matches('/'))
-                .await?;
+            let path = bucket.path().trim_start_matches('/');
+            let obj = storage_client.object().read(&host, path).await?;
             if let Some(updated) = self.last_updated {
-                // if the bucket is older than when we last checked, do nothing.
-                return Ok(updated <= obj.updated);
+                // if the object is older than the last update, do nothing
+                if obj.updated < updated {
+                    return Ok(None);
+                }
             };
-            return Ok(true);
+
+            let bytes = storage_client.object().download(&host, path).await?;
+            let contents = String::from_utf8(bytes).map_err(|e| {
+                HandlerErrorKind::General(format!("Could not read ADM Settings: {:?}", e))
+            })?;
+            let new_settings = serde_json::from_str(&contents).map_err(|e| {
+                HandlerErrorKind::General(format!("Could not read ADM Settings: {:?}", e))
+            })?;
+            return Ok(Some(new_settings));
         }
-        Ok(false)
+        Ok(None)
     }
 
-    /// Try to update the ADM filter data from the remote bucket.
-    pub async fn update(&mut self, storage_client: &cloud_storage::Client) -> HandlerResult<()> {
-        if let Some(bucket) = &self.source_url {
-            let advertiser_filters =
-                AdmFilter::advertisers_from_settings_bucket(storage_client, bucket)
-                    .await
-                    .map_err(|e| {
-                        HandlerError::internal(&format!(
-                            "Invalid bucket data in {:?}: {:?}",
-                            self.source, e
-                        ))
-                    })?;
-            self.advertiser_filters.adm_advertisers.clear();
-            for (adv, setting) in advertiser_filters.adm_advertisers {
-                self.all_include_regions.clear();
-                for country in setting.keys() {
-                    self.all_include_regions.insert(country.clone());
-                }
-                self.advertiser_filters
-                    .adm_advertisers
-                    .insert(adv.to_lowercase(), setting);
+    /// Clear and update the ADM filter data from new `AdmAdvertiserSettings`
+    pub fn update(&mut self, settings: AdmAdvertiserSettings) {
+        self.all_include_regions.clear();
+        self.advertiser_filters.adm_advertisers.clear();
+        for (adv, setting) in settings.adm_advertisers {
+            for country in setting.keys() {
+                self.all_include_regions.insert(country.clone());
             }
-            self.last_updated = Some(chrono::Utc::now());
+            self.advertiser_filters
+                .adm_advertisers
+                .insert(adv.to_lowercase(), setting);
         }
-        Ok(())
+        self.last_updated = Some(chrono::Utc::now());
```

Comment:
Instead of the current timestamp, we might want to store the last modified timestamp of the remote file to avoid some subtle ambiguities because `filter::update()` might be done significantly after `filter::fetch_new_settings()` (e.g. due to a long wait for the write lock). If there is any new update published between `filter:fetch_new_settings()` and `filter:update()`, Contile will be unable to pick it up at the next update check due to that `obj.updated < self.last_updated` check in `fetch_new_settings()`.

We're not making that much publish at the moment, so this can be handled later with a follow-up.